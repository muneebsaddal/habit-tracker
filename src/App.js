import React, { useState, useEffect } from "react";
import Header from "./components/header/Header";
import DayColumns from "./components/DayColumns";
import Habit from "./components/habit/Habit";
import AddHabitDialogue from "./components/AddHabitDialogue/AddHabitDialogue";
import HabitFormA from "./components/habitForms/formA/HabitFormA";
import HabitFormB from "./components/habitForms/formB/HabitFormB";
import uuid from 'react-uuid'

function getHabitFormData() {
	const storedData = localStorage.getItem("habitFormData");
	if (!storedData)
		return {
			name: "",
			color: "",
			question: "",
			unit: "",
			target: "",
			frequency: "",
			reminder: "",
			notes: "",
			dateChecked: [
				{
					date: "",
					checked: false,
				},
			],
		};
	return JSON.parse(storedData);
}

function getHabitArray() {
	const storedData = localStorage.getItem("habitArrayData");
	if (!storedData) return [];
	return JSON.parse(storedData);
}

function App() {
	const [addDialogueOpen, setAddDialogueOpen] = useState(false);
	const handleAddDialogueOpen = () => {
		setAddDialogueOpen(true);
	};
	const handleAddDialogueClose = () => {
		setAddDialogueOpen(false);
	};

	const [addFormOpen_A, setAddFormOpen_A] = useState(false);
	const [addFormOpen_B, setAddFormOpen_B] = useState(false);
	const handleAddFormOpen_A = () => {
		setAddFormOpen_A(true);
		handleAddDialogueClose();
		handleAddFormClose_B();
	};
	const handleAddFormOpen_B = () => {
		setAddFormOpen_B(true);
		handleAddDialogueClose();
		handleAddFormClose_A();
	};
	const handleAddFormClose_A = () => {
		setAddFormOpen_A(false);
	};
	const handleAddFormClose_B = () => {
		setAddFormOpen_B(false);
	};

	const [dialogsFormA, setDialogsFormA] = useState({
		freqDialog: false,
		reminderDialog: false,
	});

	const handleFreqDialogOpenA = () => {
		setDialogsFormA({ ...dialogsFormA, freqDialog: true });
	};
	const handleReminderDialogOpenA = () => {
		setDialogsFormA({ ...dialogsFormA, reminderDialog: true });
	};
	const handleFreqDialogCloseA = () => {
		setDialogsFormA({ ...dialogsFormA, freqDialog: false });
	};
	const handleReminderDialogCloseA = () => {
		setDialogsFormA({ ...dialogsFormA, reminderDialog: false });
	};

	const [dialogsFormB, setDialogsFormB] = useState({
		freqDialog: false,
		reminderDialog: false,
	});

	const handleFreqDialogOpenB = () => {
		setDialogsFormB({ ...dialogsFormB, freqDialog: true });
	};
	const handleReminderDialogOpenB = () => {
		setDialogsFormB({ ...dialogsFormB, reminderDialog: true });
	};
	const handleFreqDialogCloseB = () => {
		setDialogsFormB({ ...dialogsFormB, freqDialog: false });
	};
	const handleReminderDialogCloseB = () => {
		setDialogsFormB({ ...dialogsFormB, reminderDialog: false });
	};

	const [time, setTime] = useState("0:00am");

	const handleTimeChange = (data) => {
		setTime(data);
	};

	const handleTimeSubmit = (data) => {
		setDataForm((prevData) => {
			return {
				...prevData,
				reminder: data,
			};
		});
	};

	const current = new Date();

	const [dataForm, setDataForm] = useState(getHabitFormData);

	const handleChangeForm = (e) => {
		const { name, value, type, checked } = e.target;
		setDataForm((prevData) => {
			return {
				...prevData,
				[name]: type === "checkbox" ? checked : value,
			};
		});
	};

	const handleColorChanges = (color) => {
		setDataForm((prevData) => {
			return {
				...prevData,
				color: color,
			};
		});
	};

	const handleFrequencyChange = (freq) => {
		setDataForm((prevData) => {
			return {
				...prevData,
				frequency: freq,
			};
		});
	};

	useEffect(() => {
		localStorage.setItem("habitFormData", JSON.stringify(dataForm));
	}, [dataForm]);

	const [habitArray, setHabitArray] = useState(getHabitArray);

	const handleSubmitForm = (event) => {
		event.preventDefault();
		handleAddFormClose_A();
		handleAddFormClose_B();

		setHabitArray((prevState) => [
			...prevState,
			JSON.parse(localStorage.getItem("habitFormData")),
		]);
	};

	useEffect(() => {
		localStorage.setItem("habitArrayData", JSON.stringify(habitArray));
	}, [habitArray]);

	const handleHabitCheckedChange = (e) => {
		let tempState = [...habitArray];
		let tempElement = { ...tempState[0] };
		
		tempElement.dateChecked = {
			date: current,
			checked: e.target.checked,
		};
		tempState[0] = tempElement;
		setHabitArray(tempState);
	};

	console.log(habitArray);

	const habits = habitArray.map((h) => {
		return (
			<Habit
				key={JSON.stringify(h)}
				// _id={uuid()}
				habit={h}
				date={current}
				habitCheckedChange={handleHabitCheckedChange}
			/>
		);
	});

	return (
		<>
			<Header pageTitle={"Habit"} handleOpen={handleAddDialogueOpen} />
			<DayColumns currentDate={current} />
			{habits}
			<AddHabitDialogue
				open={addDialogueOpen}
				handleClose={handleAddDialogueClose}
				openFormA={handleAddFormOpen_A}
				openFormB={handleAddFormOpen_B}
			/>
			{addFormOpen_A ? (
				<HabitFormA
					openFormA={addFormOpen_A}
					handleFormClose={handleAddFormClose_A}
					handleOpenFreqDialog={handleFreqDialogOpenA}
					handleOpenReminderDialog={handleReminderDialogOpenA}
					openFreqDialog={dialogsFormA.freqDialog}
					openReminderDialog={dialogsFormA.reminderDialog}
					closeFreqDialog={handleFreqDialogCloseA}
					closeReminderDialog={handleReminderDialogCloseA}
					getTime={time}
					changeTime={handleTimeChange}
					colorChange={handleColorChanges}
					timeSubmit={handleTimeSubmit}
					formData={dataForm}
					handleFormChange={handleChangeForm}
					handleFormSubmit={handleSubmitForm}
					freqChange={handleFrequencyChange}
				/>
			) : addFormOpen_B ? (
				<HabitFormB
					openFormB={addFormOpen_B}
					handleFormClose={handleAddFormClose_B}
					handleOpenFreqDialog={handleFreqDialogOpenB}
					handleOpenReminderDialog={handleReminderDialogOpenB}
					openFreqDialog={dialogsFormB.freqDialog}
					openReminderDialog={dialogsFormB.reminderDialog}
					closeFreqDialog={handleFreqDialogCloseB}
					closeReminderDialog={handleReminderDialogCloseB}
					getTime={time}
					changeTime={handleTimeChange}
					colorChange={handleColorChanges}
					timeSubmit={handleTimeSubmit}
					formData={dataForm}
					handleFormChange={handleChangeForm}
					handleFormSubmit={handleSubmitForm}
					freqChange={handleFrequencyChange}
				/>
			) : (
				<></>
			)}
		</>
	);
}

export default App;
