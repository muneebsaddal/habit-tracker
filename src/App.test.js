import React from "react";
import { render, cleanup } from "@testing-library/react";
import App from "./App";

afterEach(cleanup);

it("should take a snapshot", () => {
	const { asFragment } = render(<App />);
	expect(asFragment(<App />)).toMatchSnapshot();
});


// import React from "react";
// import Enzyme, { shallow, render, mount } from "enzyme";
// import toJson from "enzyme-to-json";
// import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

// import App from "./App";

// Enzyme.configure({ adapter: new Adapter() });

// it("renders correctly enzyme", () => {
// 	const wrapper = shallow(<App />);

// 	expect(toJson(wrapper)).toMatchSnapshot();
// });
