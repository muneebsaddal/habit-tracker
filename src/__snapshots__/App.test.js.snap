// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <header
    class="sc-bczRLJ bvxpDg"
  >
    <h1
      class="sc-gsnTZi lkqumo"
    >
      Habit
    </h1>
    <div
      class="sc-dkzDqf eCPrbb"
    >
      <button
        class="sc-hKMtZM sc-eCYdqJ jufqKh"
      >
        <svg
          fill="currentColor"
          height="26"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="26"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 277.333H277.333V416h-42.666V277.333H96v-42.666h138.667V96h42.666v138.667H416v42.666z"
          />
        </svg>
      </button>
      <button
        class="sc-hKMtZM sc-jSMfEi jufqKh"
      >
        <svg
          fill="currentColor"
          height="25"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 3H5a1 1 0 0 0-1 1v2.59c0 .523.213 1.037.583 1.407L10 13.414V21a1.001 1.001 0 0 0 1.447.895l4-2c.339-.17.553-.516.553-.895v-5.586l5.417-5.417c.37-.37.583-.884.583-1.407V4a1 1 0 0 0-1-1zm-6.707 9.293A.996.996 0 0 0 14 13v5.382l-2 1V13a.996.996 0 0 0-.293-.707L6 6.59V5h14.001l.002 1.583-5.71 5.71z"
          />
        </svg>
      </button>
      <button
        class="sc-hKMtZM sc-gKXOVf jufqKh"
      >
        <svg
          fill="currentColor"
          height="30"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </button>
    </div>
  </header>
  <div
    class="sc-iBkjds ldSZEk"
  >
    <div
      class="sc-ftvSup kleWek"
    >
      Tue
      <br />
      10
    </div>
    <div>
      Mon
      <br />
      9
    </div>
    <div>
      Sun
      <br />
      8
    </div>
    <div>
      Sat
      <br />
      7
    </div>
    <div>
      Fri
      <br />
      6
    </div>
  </div>
  <div
    class="sc-evZas isjUon"
  />
</DocumentFragment>
`;
